src/java/org/apache/cassandra/dht/tokenallocator/IsolatedTokenAllocator.java

```diff
@@ -0,0 +1,156 @@
+
+    public static Collection<Token> allocateTokens(int additionalSplits, AbstractReplicationStrategy source)
+    {
+        Preconditions.checkArgument(additionalSplits > 0, "additionalSplits must be greater than zero");
+        Preconditions.checkNotNull(source);
+
+        List<Token> allocatedTokens = new ArrayList<>();
+        QuietAllocator allocator = new QuietAllocator(source);
+
+        // Distribute nodes among the racks in round-robin fashion in the order the user is supposed to start them.
+        var localDc = source.snitch.getLocalDatacenter();
+        // Get a list to consistently iterate over the racks as we allocate nodes
+        List<String> racks = source.getTokenMetadata().getTopology().getDatacenterRacks().get(localDc).keys().elementSet().asList();
+        int nodeId = 0;
+        int tokensAdded = 0;
+        while (tokensAdded <= additionalSplits)
+        {
+            // Allocate tokens for current node, distributing tokens round-robin among the racks.
+            var newTokens = allocator.allocateTokensForNode(nodeId, racks.get(nodeId % racks.size()));
+            tokensAdded += newTokens.size();
+            allocatedTokens.addAll(newTokens);
+            nodeId++;
+        }
+        return allocatedTokens;
+    }
+
+    /**
+     * A token allocator that takes a source token metadata and replication strategy, but clones the source token
+     * metadata with a quiet snitch, so that added nodes are not communicated to the rest of the system.
+     */
+    private static class QuietAllocator
+    {
+        private final QuietSnitch quietSnitch;
+        private final TokenMetadata quietTokenMetadata;
+        private final TokenAllocation allocation;
+        private final Map<String, SummaryStatistics> lastCheckPoint = Maps.newHashMap();
+
+        private QuietAllocator(AbstractReplicationStrategy source)
+        {
+            this.quietSnitch = new QuietSnitch();
+            this.quietTokenMetadata = source.getTokenMetadata().cloneWithNewSnitch(quietSnitch);
+            // TODO all or full replicas?
+            var rf = source.getReplicationFactor().allReplicas;
+            var numTokens = DatabaseDescriptor.getNumTokens();
+            this.allocation = TokenAllocation.create(quietSnitch, quietTokenMetadata, rf, numTokens);
+        }
+
+        private Collection<Token> allocateTokensForNode(int nodeId, String rackId)
+        {
+            // Update snitch and token metadata info to inform token allocation
+            InetAddressAndPort fakeNodeAddressAndPort = getLoopbackAddressWithPort(nodeId);
+            quietSnitch.nodeByRack.put(fakeNodeAddressAndPort, rackId);
+            quietTokenMetadata.updateTopology(fakeNodeAddressAndPort);
+
+            // Allocate tokens
+            Collection<Token> tokens = allocation.allocate(fakeNodeAddressAndPort);
+
+            // Validate ownership stats
+            validateAllocation(nodeId, rackId);
```

Comment:
The validation only serves to ensure that the allocator's understanding of the ring is the same as the node's. After the initial testing (when we are happy that we have set the allocator up correctly), we should drop the validation to save some effort.